File: lab2/Браун Дэннис/src/worker.py
#!/usr/bin/env python3
import os
import time
import signal
import sys
import json
from multiprocessing import current_process


class Worker:
    def __init__(self, worker_id, config):
        self.worker_id = worker_id
        self.config = config
        self.current_mode = "heavy"  # Начальный режим
        self.shutdown_requested = False
        self.work_params = config[f"mode_{self.current_mode}"]

        # Установка обработчиков сигналов
        signal.signal(signal.SIGTERM, self.signal_handler)
        signal.signal(signal.SIGUSR1, self.signal_handler)
        signal.signal(signal.SIGUSR2, self.signal_handler)

    def signal_handler(self, signum, frame):
        if signum == signal.SIGTERM:
            print(f"[Worker {self.worker_id}] Received SIGTERM, shutting down...")
            self.shutdown_requested = True
        elif signum == signal.SIGUSR1:
            print(f"[Worker {self.worker_id}] Switching to LIGHT mode")
            self.current_mode = "light"
            self.work_params = self.config["mode_light"]
        elif signum == signal.SIGUSR2:
            print(f"[Worker {self.worker_id}] Switching to HEAVY mode")
            self.current_mode = "heavy"
            self.work_params = self.config["mode_heavy"]

    def simulate_work(self, duration_us):
        """Имитация работы CPU"""
        end_time = time.time() + (duration_us / 1_000_000)
        while time.time() < end_time and not self.shutdown_requested:
            # Просто крутимся в цикле - это нагружает CPU
            pass

    def run(self):
        tick_count = 0
        while not self.shutdown_requested:
            # Получаем текущий CPU (если доступно)
            try:
                cpu_id = os.sched_getcpu()
            except:
                cpu_id = "N/A"

            # Выводим диагностику
            print(f"[Worker {self.worker_id}] PID: {os.getpid()}, "
                  f"Mode: {self.current_mode}, Tick: {tick_count}, "
                  f"CPU: {cpu_id}")

            # Имитация работы
            self.simulate_work(self.work_params["work_us"])

            # Пауза
            if not self.shutdown_requested:
                time.sleep(self.work_params["sleep_us"] / 1_000_000)

            tick_count += 1

        print(f"[Worker {self.worker_id}] Shutdown complete")


def main():
    if len(sys.argv) != 3:
        print("Usage: worker.py <worker_id> <config_json>")
        sys.exit(1)

    worker_id = sys.argv[1]
    config_json = sys.argv[2]

    try:
        config = json.loads(config_json)
        worker = Worker(worker_id, config)
        worker.run()
    except Exception as e:
        print(f"Worker error: {e}")
        sys.exit(1)


if __name__ == "__main__":
    main()

File: lab2/gr1sub1/KuharchukAlesia/src/worker.c
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <signal.h>
#include <time.h>
#include <sys/resource.h>
#include "config.h"

volatile sig_atomic_t worker_running = 1;
volatile sig_atomic_t current_mode = 0;
config_t worker_config;
int worker_id = 0;

void signal_handler(int sig) {
    switch(sig) {
        case SIGTERM:
            worker_running = 0;
            printf("Worker %d: Received SIGTERM, shutting down...\n", getpid());
            break;
        case SIGUSR1:
            current_mode = 1;
            printf("Worker %d: Switching to LIGHT mode\n", getpid());
            break;
        case SIGUSR2:
            current_mode = 0;
            printf("Worker %d: Switching to HEAVY mode\n", getpid());
            break;
    }
}

void setup_signals() {
    struct sigaction sa;
    sa.sa_handler = signal_handler;
    sigemptyset(&sa.sa_mask);
    sa.sa_flags = 0;
    
    sigaction(SIGTERM, &sa, NULL);
    sigaction(SIGUSR1, &sa, NULL);
    sigaction(SIGUSR2, &sa, NULL);
    signal(SIGINT, SIG_IGN);
}

void simulate_work(int work_us, int sleep_us) {
    // Разбиваем работу на мелкие части для частой проверки флага
    const int chunk_us = 1000; // 1ms chunks
    int remaining_us = work_us;
    
    while (remaining_us > 0 && worker_running) {
        int chunk = (remaining_us > chunk_us) ? chunk_us : remaining_us;
        
        clock_t start = clock();
        while (clock() < start + (chunk * CLOCKS_PER_SEC / 1000000) && worker_running) {
            // busy wait с проверкой флага
        }
        
        remaining_us -= chunk;
        
        // Частая проверка флага завершения
        if (!worker_running) break;
    }
    
    // Сон тоже с проверкой флага
    if (worker_running) {
        remaining_us = sleep_us;
        while (remaining_us > 0 && worker_running) {
            int sleep_chunk = (remaining_us > 50000) ? 50000 : remaining_us; // 50ms max
            usleep(sleep_chunk);
            remaining_us -= sleep_chunk;
        }
    }
}

int main(int argc, char *argv[]) {
    if (argc > 1) {
        worker_id = atoi(argv[1]);
    }
    
    printf("Worker[%d] %d started (parent: %d)\n", worker_id, getpid(), getppid());
    
    setup_signals();
    load_config(&worker_config);
    
    int tick_count = 0;
while (worker_running) {
    work_mode_t *mode = current_mode ? &worker_config.mode_light : &worker_config.mode_heavy;
    
    simulate_work(mode->work_us, mode->sleep_us);
    
    if (tick_count % 5 == 0) {
        print_worker_info();
    }
    
    tick_count++;
    
    // Дополнительная проверка после каждой итерации
    if (!worker_running) break;
}
    printf("Worker[%d] %d: Clean shutdown completed\n", worker_id, getpid());
    return 0;
}
